### ISSUE TITLE
Final class variable in dataclass breaks __init__ signature

### ISSUE THREAD

**Issue by ncanceill** [microsoft/pyright#9550]
**Describe the bug**
Python 3.13 allows `ClassVar[Final[<type>]]` annotations for dataclass fields. Pyright 1.1.390 accepts the annotation but gets the `__init__()` signature wrong.

When a dataclass field is annotated as `ClassVar[Final[<type>]]` it is not added as parameter to the generated `__init__()` method. The dataclass below has a final class variable `bar` with a default value followed by an instance variable `baz` without default value. Python 3.13 accepts this code and generates a method `__init__(self, baz)` as expected.

Pyright 1.1.390 seems to parse `bar` as a regular field and therefore:
- does not accept `baz` because it comes after a field with a default value
- computes a wrong signature `__init__(self, bar = 0, baz)`

**Code or Screenshots**

```python
from dataclasses import dataclass
from typing import ClassVar, Final, reveal_type

@dataclass
class Foo:
    bar: ClassVar[Final[int]] = 0
    baz: int
    # Fields without default values cannot appear after fields with default values (reportGeneralTypeIssues)

Foo(1)
# Argument missing for parameter "baz" (reportCallIssue)

reveal_type(Foo.__init__)
# Type of "Foo.__init__" is "(self: Foo, bar: int = 0, baz: int) -> None"
```

**Additional info**

This passes Pyright but forgoes the `Final` annotation:
```python
@dataclass
class Foo:
    bar: ClassVar[int] = 0
    baz: int

reveal_type(Foo.__init__)
# Type of "Foo.__init__" is "(self: Foo, baz: int) -> None"
```

This still gets the wrong signature:
```python
@dataclass
class Foo:
    baz: int
    bar: ClassVar[Final[int]] = 0

reveal_type(Foo.__init__)
# Type of "Foo.__init__" is "(self: Foo, baz: int, bar: int = 0) -> None"
```

So does this:
```python
@dataclass(kw_only=True)
class Foo:
    bar: ClassVar[Final[int]] = 0
    baz: int

reveal_type(Foo.__init__)
# Type of "Foo.__init__" is "(self: Foo, *, bar: int = 0, baz: int) -> None"
```

As noted in https://github.com/microsoft/pyright/issues/8676#issuecomment-2271844037 the Python 3.13.0 implementation does not treat dataclass fields annotated `Final[ClassVar[<type>]]` as class variables. As a result, Pyright is correct here:
```python
@dataclass
class Foo:
    bar: Final[ClassVar[int]] = 0
    baz: int
    # Fields without default values cannot appear after fields with default values (reportGeneralTypeIssues)

Foo(1)
# Argument missing for parameter "baz" (reportCallIssue)

reveal_type(Foo.__init__)
# Type of "Foo.__init__" is "(self: Foo, bar: int = 0, baz: int) -> None"
```

**VS Code extension or command-line**
Command-line, Python 3.13.0, Pyright 1.1.390

**Comment by erictraut** [comment]
Thanks for reporting the issue. I agree it's a bug.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.391.
